### PR TITLE
Invariants improvements: Adapt rate invariant for ERC20 pool with collateral precision lower than quote precision

### PR DESCRIPTION
**Pull request by grandizzy**
- if ERC20 pool and collateral precision lower than quote precision then require exchange rate to be within collateral scale
- this could happen in a take / bucket take situation when we round to collateral scale in order to be claimed
- use pending inflator to calculate increase in reserves (take penalty) on bucket take

### CODE REVIEW

File: tests/forge/invariants/base/BasicInvariants.t.sol [hunk 1]
```diff
@@ -232,13 +233,21 @@ abstract contract BasicInvariants is BaseInvariants {
                         "Exchange Rate Invariant R1, R2, R3, R4, R5, R6, R7 or R8"
                     );
                 } else {
+                    uint256 deviation = 1e13;
+
+                    // if ERC20 pool and collateral precision lower than quote precision then require exchange rate to be within collateral scale
+                    // this could happen in a take / bucket take situation when we round to collateral scale in order to be claimed
+                    if (_pool.poolType() == 0 && _quote.decimals() > IERC20Token(_pool.collateralAddress()).decimals()) {
+                        deviation = Maths.max(deviation, IERC20Pool(address(_pool)).collateralScale());
```
Comment (MikeHathaway) [inline tests/forge/invariants/base/BasicInvariants.t.sol:232]: Should this be `Maths.min` if it's supposed to be guaranteed less than that many decimals?

**Review by prat-gpt — APPROVED** [review]
LGTM!

File: tests/forge/invariants/base/BasicInvariants.t.sol (hunk 1 above)
Comment (grandizzy) [reply tests/forge/invariants/base/BasicInvariants.t.sol:232]: it actually suppose to guaranteed less than collateral scale which is calculated as `10 ** (18 - decimals)` so less precision, higher scale, hence `Maths.max` needed

Comment (MikeHathaway) [reply tests/forge/invariants/base/BasicInvariants.t.sol:232]: Right, 🤦‍♂️

**Review by MikeHathaway — APPROVED** [review]
LGTM

**Pull request merged**